docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,289 @@
+in order to increase sequence(in case of packet) or remove the commitment
+(in case of acknowledgement and timeout).
+Calling those functions implies that the application logic has successfully executed. 
+However, the handlers can return `Result` with `CodeTxBreak` after calling those methods
+which will persist the state changes that has been already done but prevent any further 
+messages to be executed in case of semantically invalid packet. In any case the 
+application modules should never return state reverting result, which will make the 
+channel unable to proceed.
+
+`ChannelKeeper.CheckOpen` method will be introduced. `ChannelChecker` function will be 
+provided by each of the `AppModule` and injected to `ChannelKeeper.Port()` at the 
+top level application. `CheckOpen` will find the correct `ChennelChecker` using the 
+`PortID` and call it, which will return an error if it is unacceptible by the application.
+
+The `ProofVerificationDecorator` will be inserted to the top level application.
+It is not safe to make each application responsible to call proof verification 
+logic, whereas application can misbehave(in terms of IBC protocol) by 
+mistake.
+
+The `ProofVerificationDecorator` should come right after the default sybil attack 
+resistent layer from the current `auth.NewAnteHandler`:
+
+```go
+// add IBC ProofVerificationDecorator to the Chain of
+func NewAnteHandler(
+  ak keeper.AccountKeeper, supplyKeeper types.SupplyKeeper, ibcKeeper ibc.Keeper,
+  sigGasConsumer SignatureVerificationGasConsumer) sdk.AnteHandler {
+  return sdk.ChainAnteDecorators(
+    NewSetUpContextDecorator(), // outermost AnteDecorator. SetUpContext must be called first
+    ...
+    NewIncrementSequenceDecorator(ak),
+    ibcante.ProofVerificationDecorator(ibcKeeper.ClientKeeper, ibcKeeper.ChannelKeeper), // innermost AnteDecorator
+  )
+}
+```
+
+The implementation of this ADR will also change the `Data` field of the `Packet` type from `[]byte` (i.e. arbitrary data) to `PacketDataI`. This also removes the `Timeout` field from the `Packet` struct. This is because the `PacketDataI` interface now contains this information. You can see details about this in [ICS04](https://github.com/cosmos/ics/tree/master/spec/ics-004-channel-and-packet-semantics#definitions). 
+
+The `PacketDataI` is the application specific interface that provides information for the execuition of the application packet. In the case of ICS20 this would be `denom`, `amount` and `address`
+
+```go
+// PacketDataI defines the standard interface for IBC packet data
+type PacketDataI interface {
+	GetCommitment() []byte // Commitment form that will be stored in the state.
+	GetTimeoutHeight() uint64
+
+	ValidateBasic() sdk.Error
+	Type() string
+}
+```
+
+Example application-side usage:
+
+```go
+type AppModule struct {}
+
+func (module AppModule) CheckChannel(portID, channelID string, channel Channel) error {
```

Comment:
Why is this a top-level function instead of a message handler?

Reply:
`CheckChannel` will be provided to `ChannelKeeper.Port()` at the top level application. Will add to the doc.